weblate/machinery/base.py

```diff
@@ -452,12 +501,17 @@ def batch_translate(self, units, user=None, threshold: int = 75):
     def _batch_translate(self, source, language, units, user=None, threshold: int = 75):
         for unit in units:
             result = unit.machinery
-            if result["best"] >= self.max_score:
+            if result is None:
+                result = unit.machinery = {}
+            elif min(*result.get("quality", (-1,))) >= self.max_score:
                 continue
-            for item in self._translate(
-                source, language, unit, user=user, threshold=threshold
-            ):
-                if result["best"] > item["quality"]:
-                    continue
-                result["best"] = item["quality"]
-                result["translation"] = item["text"]
+            translation_lists = self.translate(unit, user=user, threshold=threshold)
```

Comment:
`translate` does language mapping and usage accounting for each string, while `batch_translate` does that just once. When calling `translate` from `batch_translate`, it will do the additional language mapping for each string, and it will account the usage twice.

Adding support for translating plurals should not affect performance of non-plurals translation.

Reply:
I've refactored the code to avoid accounting for usage twice and to optimize the mapping of plurals as much as possible.